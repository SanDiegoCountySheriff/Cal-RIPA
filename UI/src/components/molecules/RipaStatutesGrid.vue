<template>
  <div>
    <v-card-title>
      <v-text-field
        v-model="search"
        append-icon="mdi-magnify"
        label="Search"
        single-line
        hide-details
      ></v-text-field>
    </v-card-title>
    <v-data-table
      :loading="loading"
      :headers="headers"
      :items="statutes"
      :search="search"
      sort-by="offenseCode"
    >
      <template v-slot:top>
        <v-toolbar flat>
          <v-toolbar-title class="tw-uppercase"
            >Admin: Maintain Statutues</v-toolbar-title
          >
          <v-spacer></v-spacer>
          <v-dialog v-model="dialog" max-width="1000px" persistent>
            <template v-slot:activator="{ on, attrs }">
              <v-btn
                color="primary"
                dark
                class="tw-mb-2"
                v-bind="attrs"
                v-on="on"
              >
                New Item
              </v-btn>
            </template>
            <v-card>
              <v-card-title>
                <span>{{ formTitle }}</span>
              </v-card-title>

              <v-card-text>
                <v-container>
                  <v-row>
                    <v-col cols="12" sm="4">
                      <v-text-field
                        v-model="editedItem.rowKey"
                        label="ID"
                        :disabled="isRowKeyDisabled"
                      ></v-text-field>
                      <v-text-field
                        v-model="editedItem.offenseCode"
                        label="Offense Code"
                      ></v-text-field>
                      <v-text-field
                        v-model="editedItem.offenseStatute"
                        label="Offense Statute"
                      ></v-text-field>
                      <v-text-field
                        v-model="editedItem.offenseValidationCD"
                        label="Offense Validation CD"
                      ></v-text-field>
                      <v-text-field
                        v-model="editedItem.offenseTxnTypeCD"
                        label="Offense Txn Type CD"
                      ></v-text-field>
                      <v-text-field
                        v-model="editedItem.offenseTypeOfStatuteCD"
                        label="Offense Type of Statute CD"
                      ></v-text-field>
                    </v-col>
                    <v-col cols="12" sm="4">
                      <v-text-field
                        v-model="editedItem.statuteLiteral"
                        label="Statute Literal"
                      ></v-text-field>
                      <v-text-field
                        v-model="editedItem.offenseDefaultTypeOfCharge"
                        label="Offense Default Type of Charge"
                      ></v-text-field>
                      <v-text-field
                        v-model="editedItem.offenseTypeOfCharge"
                        label="Offense Type of Charge"
                      ></v-text-field>
                      <v-text-field
                        v-model="editedItem.offenseLiteralIdentifierCD"
                        label="Offense Literal ID CD"
                      ></v-text-field>
                      <v-text-field
                        v-model="editedItem.offenseDegree"
                        label="Offense Degree"
                      ></v-text-field>
                    </v-col>
                    <v-col cols="12" sm="4">
                      <v-text-field
                        v-model="editedItem.bcsHierarchyCD"
                        label="BCS Hierarchy CD"
                      ></v-text-field>
                      <ripa-date-picker
                        v-model="editedItem.offenseEnacted"
                        label="Offense Enacted"
                      ></ripa-date-picker>
                      <ripa-date-picker
                        v-model="editedItem.offenseRepealed"
                        label="Offense Repealed"
                      ></ripa-date-picker>
                      <v-text-field
                        v-model="editedItem.alpsCognizantCD"
                        label="Alps Cognizant CD"
                      ></v-text-field>
                    </v-col>
                  </v-row>
                </v-container>
              </v-card-text>

              <v-card-actions>
                <v-spacer></v-spacer>
                <v-btn color="blue darken-1" text @click="close">
                  Cancel
                </v-btn>
                <v-btn
                  color="blue darken-1"
                  text
                  :disabled="isDuplicateKey"
                  @click="save"
                >
                  Save
                </v-btn>
              </v-card-actions>
            </v-card>
          </v-dialog>
          <v-dialog v-model="dialogDelete" max-width="500px" persistent>
            <v-card>
              <v-card-title
                >Are you sure you want to delete this statute?</v-card-title
              >
              <v-card-actions>
                <v-spacer></v-spacer>
                <v-btn color="blue darken-1" text @click="closeDelete"
                  >No</v-btn
                >
                <v-btn color="blue darken-1" text @click="deleteItemConfirm"
                  >Yes</v-btn
                >
                <v-spacer></v-spacer>
              </v-card-actions>
            </v-card>
          </v-dialog>
        </v-toolbar>
      </template>
      <template v-slot:item.actions="{ item }">
        <v-icon small class="tw-mr-2" @click="editItem(item)">
          mdi-pencil
        </v-icon>
        <v-icon small @click="deleteItem(item)"> mdi-delete </v-icon>
      </template>
      <template v-slot:no-data>
        <div>No Data</div>
      </template>
    </v-data-table>
  </div>
</template>

<script>
import RipaDatePicker from '@/components/atoms/RipaDatePicker'

export default {
  name: 'ripa-statutes-grid',

  components: {
    RipaDatePicker,
  },

  data() {
    return {
      search: '',
      statutes: [],
      dialog: false,
      dialogDelete: false,
      headers: [
        { text: 'ID', value: 'rowKey' },
        { text: 'Offense Code', value: 'offenseCode' },
        { text: 'Offense Statute', value: 'offenseStatute' },
        { text: 'Offense Validation CD', value: 'offenseValidationCD' },
        { text: 'Offense Txn Type CD', value: 'offenseTxnTypeCD' },
        { text: 'Offense Type of Statute CD', value: 'offenseTypeOfStatuteCD' },
        { text: 'Statute Literal', value: 'statuteLiteral' },
        {
          text: 'Offense Default Type of Charge',
          value: 'offenseDefaultTypeOfCharge',
        },
        { text: 'Offense Type of Charge', value: 'offenseTypeOfCharge' },
        { text: 'Offense Literal ID CD', value: 'offenseLiteralIdentifierCD' },
        { text: 'Offense Code', value: 'offenseDegree' },
        { text: 'BCS Hierarchy CD', value: 'bcsHierarchyCD' },
        { text: 'Offense Enacted', value: 'offenseEnacted', width: '110' },
        { text: 'Offense Repealed', value: 'offenseRepealed', width: '110' },
        { text: 'Alps Cognizant CD', value: 'alpsCognizantCD' },
        { text: 'Actions', value: 'actions', sortable: false, width: '100' },
      ],
      editedIndex: -1,
      editedItem: {
        rowKey: '',
        name: '',
        district: '',
        county: '',
      },
      defaultItem: {
        rowKey: '',
        name: '',
        district: '',
        county: '',
      },
    }
  },

  computed: {
    formTitle() {
      return this.editedIndex === -1 ? 'New Item' : 'Edit Item'
    },

    isRowKeyDisabled() {
      return this.editedIndex > -1
    },

    isDuplicateKey() {
      const filteredItems = this.statutes.filter(
        item => item.rowKey === this.editedItem.rowKey,
      )

      if (this.editedIndex === -1) {
        return filteredItems.length > 0 || this.editedItem.rowKey === ''
      }

      return false
    },
  },

  methods: {
    init() {
      this.statutes = this.items
    },

    editItem(item) {
      this.editedIndex = this.statutes.indexOf(item)
      this.editedItem = Object.assign({}, item)
      this.dialog = true
    },

    deleteItem(item) {
      this.editedIndex = this.statutes.indexOf(item)
      this.editedItem = Object.assign({}, item)
      this.dialogDelete = true
    },

    deleteItemConfirm() {
      this.statutes.splice(this.editedIndex, 1)
      if (this.onDeleteStatute) {
        this.onDeleteStatute(this.editedItem)
      }
      this.closeDelete()
    },

    close() {
      this.dialog = false
      this.$nextTick(() => {
        this.editedItem = Object.assign({}, this.defaultItem)
        this.editedIndex = -1
      })
    },

    closeDelete() {
      this.dialogDelete = false
      this.$nextTick(() => {
        this.editedItem = Object.assign({}, this.defaultItem)
        this.editedIndex = -1
      })
    },

    save() {
      if (this.editedIndex > -1) {
        Object.assign(this.statutes[this.editedIndex], this.editedItem)
      } else {
        this.statutes.push(this.editedItem)
      }

      if (this.onEditStatute) {
        this.onEditStatute(this.editedItem)
      }

      this.close()
    },
  },

  watch: {
    items(val) {
      this.statutes = val
    },
    dialog(val) {
      val || this.close()
    },
    dialogDelete(val) {
      val || this.closeDelete()
    },
  },

  created() {
    this.init()
  },

  props: {
    loading: {
      type: Boolean,
      default: false,
    },
    items: {
      type: Array,
      default: () => [],
    },
    onDeleteStatute: {
      type: Function,
      default: () => {},
    },
    onEditStatute: {
      type: Function,
      default: () => {},
    },
  },
}
</script>
