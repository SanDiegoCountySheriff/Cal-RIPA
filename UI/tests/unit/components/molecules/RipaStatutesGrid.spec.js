import RipaStatutesGrid from '@/components/molecules/RipaStatutesGrid.vue'
import { shallowMount } from '@vue/test-utils'
import Vuetify from 'vuetify'

describe('Ripa Statutes Grid', () => {
  let vuetify
  let wrapper

  beforeEach(() => {
    vuetify = new Vuetify()
  })

  afterEach(() => {
    wrapper.destroy()
  })

  const factory = propsData => {
    return shallowMount(RipaStatutesGrid, {
      vuetify,
      propsData: {
        ...propsData,
      },
    })
  }

  it('should match snapshot', () => {
    wrapper = factory()

    expect(wrapper.element).toMatchSnapshot()
  })
})
