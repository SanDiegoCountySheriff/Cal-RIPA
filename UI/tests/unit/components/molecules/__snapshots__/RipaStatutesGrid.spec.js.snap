// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Ripa Statutes Grid should match snapshot 1`] = `
<div class="ripa-statutes-grid">
  <p class="ripa-statutes-grid--info">
    Data in this table is provided by CA DOJ and maintained periodically on
    CLEW (<a href="https://clew.doj.ca.gov/sd-ab953" target="_blank" rel="noopener noreferrer">
      https://clew.doj.ca.gov/sd-ab953 </a>). Exercise caution when uploading data. Any mismatches will produce
    errors upon submission to CA DOJ.
  </p>
  <div class="v-card__title">
    <div class="v-input v-input--hide-details theme--light v-text-field v-text-field--single-line">
      <div class="v-input__control">
        <div class="v-input__slot">
          <div class="v-text-field__slot"><label for="input-2" class="v-label theme--light" style="left: 0px; position: absolute;">Search</label><input id="input-2" type="text"></div>
          <div class="v-input__append-inner">
            <div class="v-input__icon v-input__icon--append"><i aria-hidden="true" class="v-icon notranslate mdi mdi-magnify theme--light"></i></div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div class="v-data-table v-data-table--has-top v-data-table--has-bottom theme--light">
    <header class="v-sheet theme--light v-toolbar v-toolbar--flat" style="height: 64px;">
      <div class="v-toolbar__content" style="height: 64px;">
        <div class="v-toolbar__title tw-uppercase">Admin: Maintain Statutues</div>
      </div>
    </header>
    <div class="v-data-table__wrapper">
      <table>
        <colgroup>
          <col class="">
          <col class="">
          <col class="">
          <col class="">
          <col class="">
          <col class="">
          <col class="">
          <col class="">
          <col class="">
          <col class="">
          <col class="">
          <col class="">
          <col class="">
          <col class="">
          <col class="">
        </colgroup>
        <thead class="v-data-table-header">
          <tr>
            <th role="columnheader" scope="col" aria-label="ID: Not sorted. Activate to sort ascending." aria-sort="none" class="text-start sortable"><span>ID</span><i aria-hidden="true" class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light" style="font-size: 18px;"></i></th>
            <th role="columnheader" scope="col" aria-label="Offense Code: Sorted ascending. Activate to sort descending." aria-sort="ascending" class="text-start sortable active asc"><span>Offense Code</span><i aria-hidden="true" class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light" style="font-size: 18px;"></i></th>
            <th role="columnheader" scope="col" aria-label="Offense Statute: Not sorted. Activate to sort ascending." aria-sort="none" class="text-start sortable"><span>Offense Statute</span><i aria-hidden="true" class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light" style="font-size: 18px;"></i></th>
            <th role="columnheader" scope="col" aria-label="Offense Validation CD: Not sorted. Activate to sort ascending." aria-sort="none" class="text-start sortable"><span>Offense Validation CD</span><i aria-hidden="true" class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light" style="font-size: 18px;"></i></th>
            <th role="columnheader" scope="col" aria-label="Offense Txn Type CD: Not sorted. Activate to sort ascending." aria-sort="none" class="text-start sortable"><span>Offense Txn Type CD</span><i aria-hidden="true" class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light" style="font-size: 18px;"></i></th>
            <th role="columnheader" scope="col" aria-label="Offense Type of Statute CD: Not sorted. Activate to sort ascending." aria-sort="none" class="text-start sortable"><span>Offense Type of Statute CD</span><i aria-hidden="true" class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light" style="font-size: 18px;"></i></th>
            <th role="columnheader" scope="col" aria-label="Statute Literal: Not sorted. Activate to sort ascending." aria-sort="none" class="text-start sortable"><span>Statute Literal</span><i aria-hidden="true" class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light" style="font-size: 18px;"></i></th>
            <th role="columnheader" scope="col" aria-label="Offense Default Type of Charge: Not sorted. Activate to sort ascending." aria-sort="none" class="text-start sortable"><span>Offense Default Type of Charge</span><i aria-hidden="true" class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light" style="font-size: 18px;"></i></th>
            <th role="columnheader" scope="col" aria-label="Offense Type of Charge: Not sorted. Activate to sort ascending." aria-sort="none" class="text-start sortable"><span>Offense Type of Charge</span><i aria-hidden="true" class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light" style="font-size: 18px;"></i></th>
            <th role="columnheader" scope="col" aria-label="Offense Literal ID CD: Not sorted. Activate to sort ascending." aria-sort="none" class="text-start sortable"><span>Offense Literal ID CD</span><i aria-hidden="true" class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light" style="font-size: 18px;"></i></th>
            <th role="columnheader" scope="col" aria-label="Offense Code: Not sorted. Activate to sort ascending." aria-sort="none" class="text-start sortable"><span>Offense Code</span><i aria-hidden="true" class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light" style="font-size: 18px;"></i></th>
            <th role="columnheader" scope="col" aria-label="BCS Hierarchy CD: Not sorted. Activate to sort ascending." aria-sort="none" class="text-start sortable"><span>BCS Hierarchy CD</span><i aria-hidden="true" class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light" style="font-size: 18px;"></i></th>
            <th role="columnheader" scope="col" aria-label="Offense Enacted: Not sorted. Activate to sort ascending." aria-sort="none" class="text-start sortable" style="width: 110px; min-width: 110px;"><span>Offense Enacted</span><i aria-hidden="true" class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light" style="font-size: 18px;"></i></th>
            <th role="columnheader" scope="col" aria-label="Offense Repealed: Not sorted. Activate to sort ascending." aria-sort="none" class="text-start sortable" style="width: 110px; min-width: 110px;"><span>Offense Repealed</span><i aria-hidden="true" class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light" style="font-size: 18px;"></i></th>
            <th role="columnheader" scope="col" aria-label="Alps Cognizant CD: Not sorted. Activate to sort ascending." aria-sort="none" class="text-start sortable"><span>Alps Cognizant CD</span><i aria-hidden="true" class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light" style="font-size: 18px;"></i></th>
          </tr>
        </thead>
        <tbody>
          <tr class="v-data-table__empty-wrapper">
            <td colspan="15">
              <div>No Data</div>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    <div class="v-data-footer">
      <div class="v-data-footer__select">Rows per page:<div class="v-input v-input--hide-details v-input--is-label-active v-input--is-dirty theme--light v-text-field v-select">
          <div class="v-input__control">
            <div role="button" aria-haspopup="listbox" aria-expanded="false" aria-owns="list-25" class="v-input__slot">
              <div class="v-select__slot">
                <div class="v-select__selections">
                  <div class="v-select__selection v-select__selection--comma">10</div><input aria-label="Rows per page:" id="input-25" readonly="readonly" type="text" aria-readonly="false" autocomplete="off">
                </div>
                <div class="v-input__append-inner">
                  <div class="v-input__icon v-input__icon--append"><i aria-hidden="true" class="v-icon notranslate mdi mdi-menu-down theme--light"></i></div>
                </div><input type="hidden" value="10">
              </div>
              <div class="v-menu">
                <!---->
              </div>
            </div>
          </div>
        </div>
      </div>
      <div class="v-data-footer__pagination">â€“</div>
      <div class="v-data-footer__icons-before"><button type="button" disabled="disabled" class="v-btn v-btn--disabled v-btn--icon v-btn--round v-btn--text theme--light v-size--default" aria-label="Previous page"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate mdi mdi-chevron-left theme--light"></i></span></button></div>
      <div class="v-data-footer__icons-after"><button type="button" disabled="disabled" class="v-btn v-btn--disabled v-btn--icon v-btn--round v-btn--text theme--light v-size--default" aria-label="Next page"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate mdi mdi-chevron-right theme--light"></i></span></button></div>
    </div>
  </div>
</div>
`;
